rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+the value will be stats based on value derived from the function and key will be hash of file path. 
+
+For an existing table, one can use the async indexer to build the indexes concurrently with ingestion. When data is
+written:
+
+1. Use the storage strategy (hive-style, random prefix, etc.) to determine the physical location to write the data (
+   which we will do anyway for commit metadata).
+2. Update the index i.e. update MT for each index.
+    1. Synchronous update: Data table commit is not committed until all indexes are updated.
+    2. Async update: Use the [async indexer](https://hudi.apache.org/docs/metadata_indexing) to load the records
+       incrementally and update the index.
+
+#### Query Planning & Execution
+
+On receiving a query:
+
+1. Parse the query and index metadata (in `.hudi_index`) to determine the relevant functional index is available or not.
+2. Lookup (point/range) index to determine:
+    * Which files match functional index predicates
+    * Where these files are physically located (irrespective of their storage layout)
+3. Use the physical paths derived from metadata to access and read the relevant data.
+
+Below are the concrete steps on how Hudi will intercept predicate in the query and use functional index whenever it can.
+
+1. Identify the relevant functional index based on the predicate and add a new predicate based on the Functional Index
+   name if needed. Take the functional index created by
+   SQL `CREATE INDEX datestr ON hudi_table USING col_stats (DATE_FORMAT(timestamp, '%Y-%m-%d'));`
+
+- If the original data column of a functional index is used, transform the predicate to a new predicate based on the
+  Functional Index name if possible, e.g., `timestamp between 1666767600 and 1666897200`
+  to `datestr between '2022-10-26' and '2022-10-27'`;
+- If the expression of a functional index is used, replace the expression to the functional index name,
+  e.g., `DATE_FORMAT(timestamp, '%Y-%m-%d') between '2022-10-26' and '2022-10-27'`
+  to `datestr between '2022-10-26' and '2022-10-27'`;
+- If the functional index name is used in the predicate, e.g., `datestr between '2022-10-26' and '2022-10-27'`, no
+  action is needed here.
+
+2. For the functional index name appeared in the predicate, look up the corresponding index metadata from metadata table
+   and prune files that are not needed for scanning.
+
+In order to use the functional indexes, MT has to be enabled for the readers. Spark, Hive, Presto and Trino can already
+do so today. In order to generate splits, we will rely on MT. However, we will need to add some code in query engines to
+understand the predicates. For each query engine, we need to intercept the engine-specific representation of the
+predicates, and derive the relevant predicate based on the function, e.g. break down `TupleDomain` in Presto/Trino. We
+consider integrating Spark and Presto first in the implementation.
+
+#### Handling the Storage Layout Scenarios
```

Comment:
storage partitions should be agnostic to all this.

Reply:
Yes. Just mentioning here how we can handle different layouts. We can skip this section if it's too much information for the RFC, and simply mention somewhere above that partitions are at the storage level and not table level.